samza-core/src/main/java/org/apache/samza/table/remote/RemoteTableDescriptor.java

```diff
@@ -0,0 +1,239 @@
+
+import com.google.common.base.Preconditions;
+
+
+/**
+ * Table descriptor for remote store backed tables
+ *
+ * @param <K> the type of the key
+ * @param <V> the type of the value
+ */
+public class RemoteTableDescriptor<K, V> extends BaseTableDescriptor<K, V, RemoteTableDescriptor<K, V>> {
+  /**
+   * Tag to be used for provision credits for rate limiting read operations from the remote table.
+   * Caller must pre-populate the credits with this tag when specifying a custom rate limiter instance
+   * through {@link RemoteTableDescriptor#withRateLimiter(RateLimiter, CreditFunction, CreditFunction)}
+   */
+  public static final String RL_READ_TAG = "readTag";
+
+  /**
+   * Tag to be used for provision credits for rate limiting write operations into the remote table.
+   * Caller can optionally populate the credits with this tag when specifying a custom rate limiter instance
+   * through {@link RemoteTableDescriptor#withRateLimiter(RateLimiter, CreditFunction, CreditFunction)}
+   * and it needs the write functionality.
+   */
+  public static final String RL_WRITE_TAG = "writeTag";
+
+  // Input support for a specific remote store (required)
+  private TableReadFunction<K, V> readFn;
+
+  // Output support for a specific remote store (optional)
+  private TableWriteFunction<K, V> writeFn;
+
+  // Rate limiter for client-side throttling;
+  // can either be constructed indirectly from rates or overridden by withRateLimiter()
+  private RateLimiter rateLimiter;
+
+  // Rates for constructing the default rate limiter when they are non-zero
+  private int readRateLimit;
+  private int writeRateLimit;
+
+  /**
+   * Function interface for providing rate limiting credits for each table record.
+   * This interface allows callers to pass in lambda espressions which are otherwise
+   * non-serializable as-is.
+   * @param <K> the type of the key
+   * @param <V> the type of the value
+   */
+  public interface CreditFunction<K, V> extends Function<KV<K, V>, Integer>, Serializable {
+  }
+
+  private CreditFunction<K, V> readCreditFn;
+  private CreditFunction<K, V> writeCreditFn;
+
+  /**
+   * Construct a table descriptor instance
+   * @param tableId Id of the table
+   */
+  public RemoteTableDescriptor(String tableId) {
+    super(tableId);
+  }
+
+  @Override
+  public TableSpec getTableSpec() {
+    validate();
+
+    Map<String, String> tableSpecConfig = new HashMap<>();
+    generateTableSpecConfig(tableSpecConfig);
+
+    // Serialize and store reader/writer functions with config
+    tableSpecConfig.put(RemoteTableProvider.READ_FN, serializeObject("read function", readFn));
+
+    if (writeFn != null) {
+      tableSpecConfig.put(RemoteTableProvider.WRITE_FN, serializeObject("write function", writeFn));
+    }
+
+    if (this.rateLimiter == null) {
+      // Serialize and store rate limiters into config
+      Map<String, Integer> tagCreditsMap = new HashMap<>();
+
+      if (readRateLimit > 0) {
```

Comment:
Not a huge fan of this code-path - way high cyclomatic-complexity :-) 

I would instead move the map to the member variable and have it mutated at each call to `withReadLimit` and `withWriteLimit`. That would simplify things a bit IMO

Reply:
very good suggestion!